### PR TITLE
Refactor SSE

### PR DESCRIPTION
**Pull request by vladimyr**
This PR improves upon existing SSE implementation by making it _symetric_. The revised implementation is powered by two generic `SSEConnection` classes on both the server & client-side. Additionally, client-side implementation is decorated with [`debug`](https://npm.im) to make debugging connections easier.

image

### CODE REVIEW

File: client/sse.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import createDebug from 'debug';
+import { EventEmitter } from 'events';
+
+const debug = createDebug('sse-client');
+
+const Events = {
+  open: '#open'
```
Comment (vladimyr) [inline client/sse.js:7]: Native [`open`](https://developer.mozilla.org/en-US/docs/Web/API/EventSource/open_event) event is _shadowed_ by server sent `#open` in order to retrieve server-generated connection id.

File: client/sse.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+export default class SSEConnection extends EventEmitter {
+  constructor(url, options = {}) {
+    super();
+    this.options = options;
+    const { searchParams = {}, headers, timeout, ...config } = options;
+    if (debug.enabled) searchParams.debug = debug.namespace;
+    const streamUrl = createUrl(url, { searchParams });
+    this._connection = this.initialize(streamUrl, { headers, timeout, ...config });
+    this.once('open', ({ id }) => (this._id = id));
+    if (debug.enabled) this.on('message', e => debug('emitting event: %j', e));
```
Comment (vladimyr) [inline client/sse.js:16]: This kicks in once you execute the following snippet inside the browser console:
```js
localStorage.debug = 'sse-client';
```

File: client/sse.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    return connection;
+  }
+
+  close() {
+    return this._connection.close();
+  }
+
+  _emit = e => {
+    if (e.target !== this._connection) return;
+    const payload = e.data ? JSON.parse(e.data) : e;
```
Comment (vladimyr) [inline client/sse.js:54]: Local events (`open`, `error`) don't have `data` property.

File: client/sse.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    return this.emit(e.type, payload);
+  }
+
+  addListener(event, listener) {
+    const type = convertType(event);
+    super.addListener(type, listener);
+    this._connection.addEventListener(type, this._emit);
+  }
```
Comment (vladimyr) [inline client/sse.js:59]: This is event proxying/bubbling implementation from native `EventSource` to `SSEConnection` wrapper.

File: server/shared/sse/index.js [hunk 5]
```diff
@@ -37,46 +41,57 @@ class SSEConnection extends EventEmitter {
...
+    const json = JSON.stringify(data);
+    const payload = [
+      `id: ${id}`,
+      `event: ${event}`,
+      `data: ${json}`
+    ].join('\n');
+    this.write(payload);
+    if (hasProp(this.request.query, 'debug')) {
+      this.debug({ id, type: event, data });
+    }
```
Comment (vladimyr) [inline server/shared/sse/index.js:87]: In order to support client-side event logging, every event is re-emitted as a `message` with the actual event payload.

File: client/sse.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  _emit = e => {
+    if (e.target !== this._connection) return;
+    const payload = e.data ? JSON.parse(e.data) : e;
+    return this.emit(e.type, payload);
+  }
+
+  addListener(event, listener) {
+    super.addListener(event, listener);
+    this._connection.addEventListener(event, this._emit);
+  }
```
Comment (vladimyr) [inline client/sse.js:61]: This is event proxying/bubbling implementation from native `EventSource` to `SSEConnection` wrapper.

File: server/shared/sse/index.js [hunk 7]
```diff
@@ -33,50 +37,67 @@ class SSEConnection extends EventEmitter {
...
+    const json = JSON.stringify(data);
+    const payload = [
+      `id: ${id}`,
+      `event: ${event}`,
+      `data: ${json}`
+    ].join('\n');
+    this.write(payload);
+    if (hasProp(this.query, 'debug')) {
+      this.debug({ id, type: event, data });
+    }
```
Comment (vladimyr) [inline server/shared/sse/index.js:94]: In order to support client-side event logging, every event is re-emitted as a `message` with the actual event payload.

File: client/store/modules/repository/feed.js [hunk 8]
```diff
@@ -1,25 +1,62 @@
...
+    return this;
+  }
 }
 
-export {
-  feed,
-  connect
-};
+export default new RepositoryFeed({
+  baseUrl: process.env.API_PATH,
```
Comment (underscope) [inline client/store/modules/repository/feed.js:60]: Can you please update it since API_PATH no longer exists and merge the latest develop. Thanks!

**Pull request merged**